R/pivot_tables.R

```diff
@@ -357,3 +363,50 @@ ColocalizationScoresToAssay.Seurat <- function (
 
   return(object)
 }
+
+#' Utility function to fetch spatial metrics from
+#' a CellGraphAssay(5) object and convert them to a matrix
+#'
+#' @noRd
+#'
+.create_spatial_metric_matrix <- function (
+  object,
+  values_from,
+  metric,
+  ...
+) {
+  # fetch polarization scores
+  spatial_metric_table <- slot(object, name = metric)
+  if (is.null(spatial_metric_table)) {
+    abort(glue("'{metric}' scores are missing from '{class(object)}'"))
+  }
+
+  # Validate input
+  values_from <- match.arg(values_from, choices = switch(
+    metric,
+    "polarization" = c("morans_z", "morans_i"),
+    "colocalization" = c("pearson_z", "pearson", "pearson_mean",
+                         "jaccard_mean", "jaccard", "jaccard_z")
+  ))
+
+
+  pivot_assay_func <- switch(
+    metric,
+    "polarization" = PolarizationScoresToAssay,
+    "colocalization" = ColocalizationScoresToAssay
+  )
+  spatial_metric_wide_format <- pivot_assay_func(spatial_metric_table, values_from, ...)
+
+  # Make sure that the components match
+  # Create an empty matrix (all 0's)
+  tofillMat <- matrix(data = 0,
+                      nrow = nrow(spatial_metric_wide_format),
+                      ncol = ncol(object),
+                      dimnames = list(rownames(spatial_metric_wide_format), colnames(object)))
+
+  # Fill matrix where it overlaps
+  # Any missing columns will be kept as 0's
```

Comment:
Zero feels like a decent default, but in actuality it is a missing value. Should the user be able to control the fill value?

Reply:
The problem is that (unless this has been changed in Seurat 5) Assays do not allow missing values. The idea with these pivot functions was to create assays that can be stored in Seurat objects, so we have to set these values to 0. This was one of the concerns I had with this conversion, but I still think that it is quite useful to be able to store the spatial metrics in assays. It would get even worse if we decide to filter spatial metrics based on marker counts. Not sure what to do about it though :-/

Reply:
Okay, that's a fair point! But for example the `counts` slot is a sparse matrix? But the missing values are interpreted as zeros. 

I suppose the issue arises when e.g. PCA is applied on the polarization matrix. One would need to use a sparse versions of algorithms and interpret zeros as missing values to avoid bias from zeros. It seems quite farfetched at the current time.

Reply:
Yeah I think you're right that certain methods might be unfit to run on the spatial metrics as they are represented now. Missing values is a bit of a pin :-/

Reply:
Yeah, but we'll get to that in time! It is almost a philosophical question how to represent the position of something that doesn't exist, haha. Maybe it is zero...?